Support for Redis 6 ACLs

michael-grunder opened the pull request:
Adds support for ACLs in `Redis`, `RedisCluster`, and `RedisArray` classes.

Additionally, it allows users to customize how they group persistent connections (e.g. they may configure phpredis to add `persistent_id`, `user`, or hashed password to the hash key.  The default behavior leaves the string unchanged from the current released version of phpredis.

michael-grunder committed:
WIP:  Initial additon of 'user' Redis classes

* Adds `user` to RedisSock so we can authenticate either in the old
  way just with password, or with username and password.

* Refactored lots of boilerplate code to load configuration information
  from hash tables, and replaced PHP5 conversion logic with PHP7's
  zval_get_<type> helpers.

michael-grunder committed:
WIP:  More work on ACL support.

* Implements 'nice' handlers for the ACL (e.g. turning maps into
  PHP associative arrays).

* Initial implementation of the ACL command in Redis Cluster.

* Some minor refactoring of the auth logic in RedisSock

michael-grunder committed:
WIP:  Custom handler for ACL log

michael-grunder committed:
WIP:  ACLs

Refactor `ACL` command handler and implement it for RedisCluster.

michael-grunder committed:
WIP:  ACLs

Revert to previous auth argument and either treat it as a password
if it's a scalar value, or as a username/password if it's an array
with two elements.

A couple of leak fixes

michael-grunder committed:
WIP:  ACls

Formalize how we treat the $auth argument and add the ability to
specify a username and password for our unit tests.

michael-grunder committed:
WIP:  ACLs

Adds configurable port for Redis unit tests.  Strictly speaking this
isn't ACL related but it makes it a lot easier to switch between ACL
enabled and disabled instances to test the ACL functionality.

michael-grunder committed:
WIP: ACLs - Fix auth fragment construction

michael-grunder committed:
WIP:  ACLs - Update unit test logic

* Update session test logic to handle username/password if provided.

* Fix our throwable error handler to throw an exception for the new
  'ERR AUTH' error type.

michael-grunder committed:
WIP:  ACls

Make sure we return a copy of the password when set.

michael-grunder committed:
WIP:  ACLs

* Fix auth args for RedisArray tests

michael-grunder committed:
Maverick attempt at building/installing Redis 6

michael-grunder committed:
Need sudo for make install

michael-grunder committed:
Back out of the redis directory

michael-grunder committed:
WIP:  ACLs

* Use 'ACL SETUSER' instead of 'ACL ADDUSER' since ADDUSER doesn't
  exist.

michael-grunder committed:
WIP:  ACLs

Use a Redis 6 ppa instead of building redis 6 from source.

michael-grunder committed:
WIP:  ACLs

Make sure we add the repository noninteractively

michael-grunder committed:
WIP:  ACLs

Add some convenience unit test assertions and round out the rest
of our tests for the ACL command.

michael-grunder committed:
WIP: ACLs

Adds an exception matching assertion helper and tests to make sure
AUTH errors like NOPERM and WRONGPASS throw exceptions as opposed
to simply returning `false` and setting `getLastError()`.

michael-grunder committed:
WIP:  ACLs

Fix noperm login to use a user we actually created

michael-grunder committed:
WIP:  ACLs - User user and password for unit tests.

michael-grunder committed:
WIP:  ACLs - backward compatibility and RedisArray fixes

We probably can't change `getAuth()` at all so make the return value
dynamic, depending on the auth provided to us.

Update user/pass handling in RedisArray so it's exactly the same as
Redis and RedisCluster.

michael-grunder committed:
WIP:  ACLs

Refactor ra_load_array to get rid of a bunch of ancient boilerplate
code and update the logic so the `auth` argument has identical
semantics to every other auth argument (takes either a string or an
array with one or two elements).

michael-grunder committed:
WIP:  ACLs - Allow auth during `pconnect`

Allow the user to specify password or user/password during
`connect` and `pconnect` to prepare for user juggling with
persistent connections.

michael-grunder committed:
WIP:  ACLs

Refactor liveness checking logic.

michael-grunder committed:
WIP:  ACLs - pconnect pooling and refactoring

* Add username to the persistent connection pool hash key so
  persistent connections aren't shared between multiple logins.

* Refactor registering persistent resources and move #if directives
  to a single place.

michael-grunder committed:
WIP:  ACLs

Update our `make-cluster.sh` shell script to make it easier to
spin up and down a cluster with ACL authentication enabled.

michael-grunder committed:
ACLS:  Refactor persistent list storage

michael-grunder committed:
WIP:  ACLs

This commit adds a final piece to ACL handling by adding
functionality and configuration to persistent connections.

The commit adds a new ini setting `redis.pconnect.pool_pattern`
that allows admins to configure what elements they wish to include
when constructing the hash key for pooled connections.

Presently this key takes the form 'phpredis_<host>:<port>', which
continues to be the case by default.

However, additional specifiers may be passed to group connections
in different ways:

FMT Meaning
--- -------------------------------------------------
p   Adds the user supplied 'persistent_id' to the key
u   Adds redis username (must be supplied on connect)
a   sha256(salt + username + password).  This would
    allow users to group persistent connections by
    authenticated user.

michael-grunder committed:
WIP:  ACLs - rework pool id printf method

Change pool format specifiers to give users more flexibility

'i' - persistent_id (if set)
'u' - user name (if set)
'p' - hashed password (if set).  Redis accounts may have multiple valid
      passwords so this would allow keeping those accounts separate.
'a' - hashed user + password combination.  When used 'u' is redundant.

michael-grunder committed:
WIP:  ACLs - Don't use smart_str_append_printf

michael-grunder committed:
WIP: ACLs - Tiny whitespace fix

nicolasff commented on library.c:
```diff
@@ -1575,6 +1794,96 @@ redis_xinfo_reply(INTERNAL_FUNCTION_PARAMETERS, RedisSock *redis_sock, zval *z_t
+        if (!(key = redis_sock_read(redis_sock, &klen)) ||
+            redis_read_reply_type(redis_sock, &type, &vlen) < 0 ||
+            (type != TYPE_BULK && type != TYPE_MULTIBULK) ||
+            klen > INT_MAX || vlen > INT_MAX)
```
`klen` can't be greater than `INT_MAX` since it's an `int`.

michael-grunder replied:
Good catch! :man_facepalming:

nicolasff commented on redis_commands.c:
```diff
@@ -2047,26 +2047,51 @@ int redis_pfcount_cmd(INTERNAL_FUNCTION_PARAMETERS, RedisSock *redis_sock,
-    // Construct our AUTH command
-    *cmd_len = REDIS_CMD_SPPRINTF(cmd, "AUTH", "s", pw, pw_len);
+    /* We weren't provided user or password, short circuit */
+    if (!user && !pass)
```
should this be `||` instead of `&&`?

michael-grunder replied:
No, it's ok if they only send `pass`.

That would have us `AUTH <pass>` (backward compatible)
When we have `user` and `pass` it will send `AUTH <user> <pass>` (new Redis 6 feature)

That said it might be better to just add a third `else` and return FAILURE so the logic is easier to follow.

nicolasff reviewed (commented):
Impressive change! I didn't look too closely or anything but from just a quick scan a couple of things jumped at me.

michael-grunder committed:
ACLs: int can't be > INT_MAX by definition

michael-grunder committed:
WIP:  ACLs - Explicitly abort early with wrong user/pass combination

michael-grunder committed:
WIP: ACLs - The else if is redundant

`pass` must be non null on this else branch as we've covered
every other possibility.

michael-grunder committed:
WIP: ACLs - Minor cleanup, fix tests.

* I guess it's theoretically possible that PHP doesn't have 'ctype_xdigit'
so implement it ourselves if that is the case.

* Remove unused 'php_password.h' include in library.c

michael-grunder committed:
Don't run the TLS test if we don't see Redis on 6378

michael-grunder committed:
ACLs:  Protect against invalid auth args + tests

michael-grunder committed:
WIP:  ACLs - Clean up loading clusters from php.ini

michael-grunder committed:
WIP:  ACLs - formalize user/pass extraction.

Simplify success/failure logic for auth and nail down argument semantics
for AUTH extraction.

This will allow the user to AUTH with just a password like so:

$obj_r->auth([NULL, 'password']), or $obj_r->auth([false, 'password']);

michael-grunder committed:
WIP:  ACLs - Use identical auth semantics everywhere

michael-grunder committed:
ACLs:  Use accessor to set credentials

yatsukhnenko commented on library.c:
```diff
@@ -129,34 +148,110 @@ static int reselect_db(RedisSock *redis_sock) {
+
+    if (redis_sock->user)
+        redis_cmd_append_sstr_zstr(str, redis_sock->user);
+    if (redis_sock->pass)
```
You have already checked pass

yatsukhnenko commented on library.c:
```diff
@@ -129,34 +148,110 @@ static int reselect_db(RedisSock *redis_sock) {
+    if (redis_sock->pass)
+        zend_string_release(redis_sock->pass);
+
+    redis_sock->user = NULL;
```
Move this two lines to appropriate `if` statement

yatsukhnenko commented on library.c:
```diff
@@ -129,34 +148,110 @@ static int reselect_db(RedisSock *redis_sock) {
+    if (redis_sock->pass == NULL)
+        return NULL;
+
+    if (redis_sock->user && redis_sock->pass) {
```
Pass already checked

yatsukhnenko commented on library.c:
```diff
@@ -594,6 +708,106 @@ union resparg {
+    ops->hash_init(ctx);
+    ops->hash_update(ctx, (const unsigned char *)ZSTR_VAL(salted.s), ZSTR_LEN(salted.s));
+
+    digest = zend_string_alloc(ops->digest_size, 0);
```
You use `digest` as `char *` so zend_string is unnecessary and `emalloc` should be fine

yatsukhnenko commented on library.c:
```diff
@@ -1575,6 +1793,96 @@ redis_xinfo_reply(INTERNAL_FUNCTION_PARAMETERS, RedisSock *redis_sock, zval *z_t
+    }
+
+    if (IS_ATOMIC(redis_sock)) {
+        RETVAL_ZVAL(&zret, 0, 0);
```
`zret` is uninitialised here if `redis_read_reply_type` fails

yatsukhnenko commented on library.c:
```diff
@@ -1798,65 +2104,45 @@ redis_sock_create(char *host, int host_len, int port,
-    redis_sock->err = NULL;
 
     redis_sock->scan = 0;
-
```
You removed all `NULL` initializers, maybe all `0` are redundant and should be deleted too?

yatsukhnenko commented on library.c:
```diff
@@ -2975,4 +3307,149 @@ redis_read_variant_reply_strings(INTERNAL_FUNCTION_PARAMETERS, RedisSock *redis_
+    }
+
+    if (num == 2) {
+        if ((ztmp = zend_hash_index_find(ht, 0)))
```
Don't you want to make associative array?

yatsukhnenko commented on tests/RedisTest.php:
```diff
@@ -53,19 +51,51 @@ protected function mstime() {
+        $this->getAuthParts($user, $pass);
+
+        if ($user && $pass) {
+            return "auth[]=$user&auth[]=$pass";
```
It would be great if users can set `auth[pass]=$pass&auth[user]=$user` and don't think about ordering

yatsukhnenko reviewed (approved):
There are only code style comments. All bugs went unnoticed 😅

michael-grunder committed:
ACLs:  Review changes
